rfcs/20200306-single-client-parameter-server.md

```diff
@@ -0,0 +1,517 @@
+
+
+###### When materialing a `Future` object
+
+It is possible that a function is executed but its corresponding worker fails when users try to consume its output. In this case, we will give users a `None` value and set an error in the `Future` object.
+
+We can mitigate the problem by eagerly materializing function outputs when they are passed to `local_results`. 
+
+We can explore mechanisms to recover these objects in the future. In the short-term, users can choose to write the results to variables on parameter servers, just like a Keras metric.
+
+
+##### Parameter server failure
+
+When a parameter server fails, the error will be propagated to the client via workers. Since the latest values of variables on the failed parameter servers are gone, there is no way for the client to recover them. Therefore the training will pause until the failed parameter server is back. The client then needs to clean up other variables on other parameter servers, rebuild all the variables and load variable values from a checkpoint. To trigger this process, the simplest method is to restart the client as well.
+
+
+##### Client failure
+
+When a client fails, some scheduled functions will continue to run on workers. No new functions will be scheduled. When the client comes back, it will create variables, load from a checkpoint, schedule functions with a new context id. All the old variables will be garbage-collected when we reset their eager contexts.
+
+#### Resource Management for Workers
+
+When a worker has recovered from failure, we will need to rebuild iterators, worker-local variables, lookup tables and other resources on that worker that don’t need to be read from a checkpoint. This means that the client will have to keep track of these iterators, worker-local variables and other resources.
+
+Keeping track of resources and rebuilding them will be achieved depending how users create their resources:
+
+
+*   we will record iterators created via `tf.distribute`’s API; The state of a rebuilt iterator will be lost. We can recover their states as future work.
+*   we will capture the creation of worker-local variables via variable creator scopes.
+*   in the future we will provide users an API to create worker-local resources. We will capture these resources in the API.
+
+If users create iterators or other resources inside a function but don’t expose them as outputs, we will not rebuild them.
+
+
+#### The Unknown of Scheduled Functions
+
+For functions that have been scheduled, it is difficult for the client to know whether they have actually been executed or not when the client detects their corresponding worker failure. Therefore, in addition to inform users of this uncertainty in the case of worker failure, we should do the following to reduce this uncertainty:
```

Comment:
Depending on the consistency guarantees we make for the parameter server:
Would it be possible (or even make sense) for the parameter server to track which functions have and haven't been completed?

This may let us do things like know exactly which functions need to be 
re-executed if a parameter server fails and a checkpoint needs to be reloaded.

Then again, if people are training asynchronously I guess it's not necessarily worth trying to provide reproducibility guarantees, or guarantees that each step will be run exactly once.

Reply:
Our current design allows users to receive an object from `strategy.schedule()`, and when calling `local_results(object)` one gets the result. If it's not executed or not fully executed due to worker preemption, the strategy will find another worker to take over. For other kinds of errors, it is possible to design it in a way that such error would be discovered upon calling `strategy.local_results()` or `strategy.join()`. For the case where PS fails, the whole cluster would be restarted and a checkpoint previously saved is loaded.